internal/dns/dns.go

```diff
@@ -141,14 +181,80 @@ func (h *handler) GetDNSDomain(clusterName, baseDNSDomainName string) (*DNSDomai
 	}, nil
 }
 
+// ValidateBaseDNS validates the specified base domain name
 func (h *handler) ValidateBaseDNS(domain *DNSDomain) error {
-	return validations.ValidateBaseDNS(domain.Name, domain.ID, domain.Provider)
+	dnsProvider := h.providerFactory.GetProvider(domain)
+	dnsNameFromService, err := dnsProvider.GetDomainName()
+	if err != nil {
+		return errors.Errorf("Can't validate base DNS domain: %v", err)
+	}
+
+	dnsNameFromCluster := strings.TrimSuffix(domain.Name, ".")
+	if dnsNameFromService == dnsNameFromCluster {
+		// Valid domain
+		return nil
+	}
+	if matched, _ := regexp.MatchString(".*\\."+dnsNameFromService, dnsNameFromCluster); !matched {
+		return errors.Errorf("Domain name isn't correlated properly to DNS service")
+	}
+	return nil
 }
 
-func (b *handler) ValidateDNSRecords(cluster common.Cluster, domain *DNSDomain) error {
+func (h *handler) ValidateDNSRecords(cluster common.Cluster, domain *DNSDomain) error {
 	vipAddresses := []string{domain.APIDomainName, domain.IngressDomainName}
 	if swag.StringValue(cluster.HighAvailabilityMode) == models.ClusterHighAvailabilityModeNone {
 		vipAddresses = append(vipAddresses, domain.APIINTDomainName)
 	}
-	return validations.CheckDNSRecordsExistence(vipAddresses, domain.ID, domain.Provider)
+	if err := h.checkDNSRecordsExists(vipAddresses, domain, "A"); err != nil {
+		return err
+	}
+	if err := h.checkDNSRecordsExists(vipAddresses, domain, "AAAA"); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (h *handler) checkDNSRecordsExists(names []string, domain *DNSDomain, recordType string) error {
+	dnsProvider := h.providerFactory.GetProviderByRecordType(domain, recordType)
+	if dnsProvider == nil {
+		return nil
+	}
+	for _, name := range names {
+		res, err := dnsProvider.GetRecordSet(name)
+		if err != nil {
+			return errors.Errorf("Can't verify DNS record set existence: %v", err)
+		}
+		if res != "" {
+			return errors.Errorf("DNS domain already exists")
+		}
+	}
+	return nil
+}
+
+func (f *defaultDNSProviderFactory) GetProviderByRecordType(domain *DNSDomain, recordType string) dnsproviders.Provider {
+	switch domain.Provider {
+	case "route53":
+		return dnsproviders.Route53{
+			RecordSet: dnsproviders.RecordSet{
+				RecordSetType: recordType,
+				TTL:           60,
+			},
+			HostedZoneID: domain.ID,
+			SharedCreds:  true,
+		}
+	}
+	f.log.Debugf("No suitable implementation for DNS provider %s", domain.Provider)
+	return nil
+}
+
+func (f *defaultDNSProviderFactory) GetProvider(domain *DNSDomain) dnsproviders.Provider {
+	switch domain.Provider {
+	case "route53":
+		return dnsproviders.Route53{
+			HostedZoneID: domain.ID,
+			SharedCreds:  true,
+		}
+	}
+	f.log.Debugf("No suitable implementation for DNS provider %s", domain.Provider)
```

Comment:
Why do we need this if there is only one provider (route53)?

Reply:
@danielerez?

Reply:
The switch was added just for future expandability, i.e. support a new provider. Can be refactored though if not needed, but probably out of the scope of this PR.